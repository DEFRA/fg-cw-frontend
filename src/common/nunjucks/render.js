import camelCase from "lodash/camelCase.js";
import { environment } from "./nunjucks.js";

/**
 * @param {string} name
 * @param {object} params
 * @param {string} [block]
 */
export const render = (name, params, block) => {
  const macro = {
    path: `${name}/macro.njk`,
    name: camelCase(name.split("-")),
    params: JSON.stringify(params, null, 2),
  };

  const callComponent = block
    ? `{%- call ${macro.name}(${macro.params}) -%}${block}{%- endcall -%}`
    : `{{- ${macro.name}(${macro.params}) -}}`;

  return environment.renderString(
    `{%- from "${macro.path}" import ${macro.name} -%}
    ${callComponent}
    `,
    {},
  );
};
