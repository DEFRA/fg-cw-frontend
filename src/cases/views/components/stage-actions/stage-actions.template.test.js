import { describe, expect, test } from "vitest";
import { render } from "../../../../common/nunjucks/render.js";

describe("stage-actions", () => {
  test("renders basic stage actions", () => {
    const component = render("stage-actions", {
      caseId: "123",
      stage: {
        actions: {
          idPrefix: "actionId",
          name: "actionId",
          legend: "Decision",
          items: [
            {
              value: "approve",
              text: "Approve",
              checked: false,
            },
            {
              value: "reject",
              text: "Reject",
              checked: false,
            },
          ],
        },
        saveDisabled: false,
      },
    });

    expect(component).toMatchSnapshot();
  });

  test("renders stage actions with conditional textarea", () => {
    const component = render("stage-actions", {
      caseId: "456",
      stage: {
        actions: {
          idPrefix: "actionId",
          name: "actionId",
          legend: "Decision",
          errorMessage: null,
          items: [
            {
              value: "approve",
              text: "Approve",
              checked: true,
              conditional: {
                id: "approve-comment",
                name: "approve-comment",
                value: "This looks good",
                label: { text: "Approval reason" },
                hint: { text: "Please provide a reason" },
                rows: 3,
                required: true,
              },
            },
            {
              value: "reject",
              text: "Reject",
              checked: false,
            },
          ],
        },
        saveDisabled: false,
      },
    });

    expect(component).toMatchSnapshot();
  });

  test("renders with error message", () => {
    const component = render("stage-actions", {
      caseId: "789",
      stage: {
        actions: {
          idPrefix: "actionId",
          name: "actionId",
          legend: "Decision",
          errorMessage: {
            text: "Please select an action",
            href: "#actionId",
          },
          items: [
            {
              value: "approve",
              text: "Approve",
              checked: false,
            },
          ],
        },
        saveDisabled: false,
      },
    });

    expect(component).toMatchSnapshot();
  });

  test("renders with save button disabled", () => {
    const component = render("stage-actions", {
      caseId: "999",
      stage: {
        actions: {
          idPrefix: "actionId",
          name: "actionId",
          legend: "Decision",
          items: [
            {
              value: "approve",
              text: "Approve",
              checked: false,
            },
          ],
        },
        saveDisabled: true,
      },
    });

    expect(component).toMatchSnapshot();
  });

  test("renders multiple actions with mixed conditionals", () => {
    const component = render("stage-actions", {
      caseId: "mixed-123",
      stage: {
        actions: {
          idPrefix: "actionId",
          name: "actionId",
          legend: "What do you want to do?",
          items: [
            {
              value: "approve",
              text: "Approve",
              checked: false,
              conditional: {
                id: "approve-comment",
                name: "approve-comment",
                value: "",
                label: { text: "Approval notes" },
                rows: 3,
                required: true,
              },
            },
            {
              value: "on-hold",
              text: "Put on hold",
              checked: true,
              conditional: {
                id: "on-hold-comment",
                name: "on-hold-comment",
                value: "Waiting for more information",
                label: { text: "Reason for hold" },
                hint: { text: "Optional notes" },
                rows: 2,
                required: false,
              },
            },
            {
              value: "reject",
              text: "Reject",
              checked: false,
            },
          ],
        },
        saveDisabled: false,
      },
    });

    expect(component).toMatchSnapshot();
  });

  test("renders with textarea error", () => {
    const component = render("stage-actions", {
      caseId: "error-case",
      stage: {
        actions: {
          idPrefix: "actionId",
          name: "actionId",
          legend: "Decision",
          items: [
            {
              value: "approve",
              text: "Approve",
              checked: true,
              conditional: {
                id: "approve-comment",
                name: "approve-comment",
                value: "",
                label: { text: "Approval reason" },
                rows: 3,
                required: true,
                errorMessage: {
                  text: "Approval reason is required",
                  href: "#approve-comment",
                },
              },
            },
          ],
        },
        saveDisabled: false,
      },
    });

    expect(component).toMatchSnapshot();
  });
});
