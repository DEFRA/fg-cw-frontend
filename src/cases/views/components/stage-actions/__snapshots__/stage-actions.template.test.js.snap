// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`stage-actions > renders basic stage actions 1`] = `
"
<form method="POST" action="/cases/123/stage/outcome">

  <div class="govuk-form-group">
  <fieldset class="govuk-fieldset">
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--m">
      Decision
    </legend>
    <div class="govuk-radios" data-module="govuk-radios">
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="actionId" name="actionId" type="radio" value="approve">
        <label class="govuk-label govuk-radios__label" for="actionId">
          Approve
        </label>
      </div>
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="actionId-2" name="actionId" type="radio" value="reject">
        <label class="govuk-label govuk-radios__label" for="actionId-2">
          Reject
        </label>
      </div>
    </div>
  </fieldset>
</div>


  <button type="submit" class="govuk-button" data-module="govuk-button">
  Save
</button>
</form>
"
`;

exports[`stage-actions > renders multiple actions with mixed conditionals 1`] = `
"
<form method="POST" action="/cases/mixed-123/stage/outcome">

  <div class="govuk-form-group">
  <fieldset class="govuk-fieldset">
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--m">
      What do you want to do?
    </legend>
    <div class="govuk-radios" data-module="govuk-radios">
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="actionId" name="actionId" type="radio" value="approve" data-aria-controls="conditional-actionId">
        <label class="govuk-label govuk-radios__label" for="actionId">
          Approve
        </label>
      </div>
      <div class="govuk-radios__conditional govuk-radios__conditional--hidden" id="conditional-actionId">
        <div class="govuk-form-group">
    <label class="govuk-label" for="approve-comment">
      Approval notes
    </label>
    <textarea class="govuk-textarea" id="approve-comment" name="approve-comment" rows="3"></textarea>
  </div>
      </div>
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="actionId-2" name="actionId" type="radio" value="on-hold" checked data-aria-controls="conditional-actionId-2">
        <label class="govuk-label govuk-radios__label" for="actionId-2">
          Put on hold
        </label>
      </div>
      <div class="govuk-radios__conditional" id="conditional-actionId-2">
        <div class="govuk-form-group">
    <label class="govuk-label" for="on-hold-comment">
      Reason for hold
    </label>
    <div id="on-hold-comment-hint" class="govuk-hint">
      Optional notes
    </div>
    <textarea class="govuk-textarea" id="on-hold-comment" name="on-hold-comment" rows="2" aria-describedby="on-hold-comment-hint">Waiting for more information</textarea>
  </div>
      </div>
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="actionId-3" name="actionId" type="radio" value="reject">
        <label class="govuk-label govuk-radios__label" for="actionId-3">
          Reject
        </label>
      </div>
    </div>
  </fieldset>
</div>


  <button type="submit" class="govuk-button" data-module="govuk-button">
  Save
</button>
</form>
"
`;

exports[`stage-actions > renders stage actions with conditional textarea 1`] = `
"
<form method="POST" action="/cases/456/stage/outcome">

  <div class="govuk-form-group">
  <fieldset class="govuk-fieldset">
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--m">
      Decision
    </legend>
    <div class="govuk-radios" data-module="govuk-radios">
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="actionId" name="actionId" type="radio" value="approve" checked data-aria-controls="conditional-actionId">
        <label class="govuk-label govuk-radios__label" for="actionId">
          Approve
        </label>
      </div>
      <div class="govuk-radios__conditional" id="conditional-actionId">
        <div class="govuk-form-group">
    <label class="govuk-label" for="approve-comment">
      Approval reason
    </label>
    <div id="approve-comment-hint" class="govuk-hint">
      Please provide a reason
    </div>
    <textarea class="govuk-textarea" id="approve-comment" name="approve-comment" rows="3" aria-describedby="approve-comment-hint">This looks good</textarea>
  </div>
      </div>
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="actionId-2" name="actionId" type="radio" value="reject">
        <label class="govuk-label govuk-radios__label" for="actionId-2">
          Reject
        </label>
      </div>
    </div>
  </fieldset>
</div>


  <button type="submit" class="govuk-button" data-module="govuk-button">
  Save
</button>
</form>
"
`;

exports[`stage-actions > renders with error message 1`] = `
"
<form method="POST" action="/cases/789/stage/outcome">

  <div class="govuk-form-group govuk-form-group--error">
  <fieldset class="govuk-fieldset" aria-describedby="actionId-error">
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--m">
      Decision
    </legend>
    <p id="actionId-error" class="govuk-error-message">
      <span class="govuk-visually-hidden">Error:</span> Please select an action
    </p>
    <div class="govuk-radios" data-module="govuk-radios">
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="actionId" name="actionId" type="radio" value="approve">
        <label class="govuk-label govuk-radios__label" for="actionId">
          Approve
        </label>
      </div>
    </div>
  </fieldset>
</div>


  <button type="submit" class="govuk-button" data-module="govuk-button">
  Save
</button>
</form>
"
`;

exports[`stage-actions > renders with save button disabled 1`] = `
"
<form method="POST" action="/cases/999/stage/outcome">

  <div class="govuk-form-group">
  <fieldset class="govuk-fieldset">
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--m">
      Decision
    </legend>
    <div class="govuk-radios" data-module="govuk-radios">
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="actionId" name="actionId" type="radio" value="approve">
        <label class="govuk-label govuk-radios__label" for="actionId">
          Approve
        </label>
      </div>
    </div>
  </fieldset>
</div>


  <button type="submit" disabled aria-disabled="true" class="govuk-button" data-module="govuk-button">
  Save
</button>
</form>
"
`;

exports[`stage-actions > renders with textarea error 1`] = `
"
<form method="POST" action="/cases/error-case/stage/outcome">

  <div class="govuk-form-group">
  <fieldset class="govuk-fieldset">
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--m">
      Decision
    </legend>
    <div class="govuk-radios" data-module="govuk-radios">
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="actionId" name="actionId" type="radio" value="approve" checked data-aria-controls="conditional-actionId">
        <label class="govuk-label govuk-radios__label" for="actionId">
          Approve
        </label>
      </div>
      <div class="govuk-radios__conditional" id="conditional-actionId">
        <div class="govuk-form-group govuk-form-group--error">
    <label class="govuk-label" for="approve-comment">
      Approval reason
    </label>
    <p id="approve-comment-error" class="govuk-error-message">
      <span class="govuk-visually-hidden">Error:</span> Approval reason is required
    </p>
    <textarea class="govuk-textarea govuk-textarea--error" id="approve-comment" name="approve-comment" rows="3" aria-describedby="approve-comment-error"></textarea>
  </div>
      </div>
    </div>
  </fieldset>
</div>


  <button type="submit" class="govuk-button" data-module="govuk-button">
  Save
</button>
</form>
"
`;
